### PR TITLE
buffer implementation

### LINKED ISSUE #119: Event 'connected' is not called.
In my snippet below the 'connected' event is never fired, even though the logging works normally. In debugger I have verified that this line is executed: https://github.com/logentries/le_node/blob/master/src/node_modules/logger.js#L199, but `this` here is not the Logger object but some other EventEmitter.

```
var logger = new Logentries(config.logentries);
logger.on('connected', function () {
    logger.lastConnect = new Date();
});
```

### PR DESCRIPTION
**Pull request by stopal-r7**
circular buffer implementation.
bufferSize defaults to 16192 now.
removal of 'buffer full' error message and event.
fix for non exposed connection events.

this should fix #119 and #110(main problem was already fixed with connection refactoring)

### CODE REVIEW

**Comment by stopal-r7** [comment]
to have an idea: @pclune-r7 
cc @dtracey-r7

File: src/node_modules/logger.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ import {build}                              from 'serialize';
 import {Writable}                           from 'stream';
 import {BadOptionsError, LogentriesError}   from 'error';
 import reconnect_core                       from 'reconnect-core';
+import {RingBuffer}							from 'ringbuffer';
```
Comment (gyakushev) [inline src/node_modules/logger.js]: nit: indent

File: src/node_modules/logger.js [hunk 2]
```diff
@@ -22,7 +23,8 @@ const $reconnect                = Symbol('reconnect module');
 const $reconnectInitialDelay    = Symbol('reconnectionInitialDelay property');
 const $reconnectMaxDelay        = Symbol('reconnectionMaxDelay property');
 const $reconnectBackoffStrategy = Symbol('reconnectBackoffStrategy property');
-const $bufferFull               = Symbol('bufferFull property');
+const $bufferWasFull			= Symbol('bufferFull property');
```
Comment (gyakushev) [inline src/node_modules/logger.js]: nit: indent

File: src/node_modules/logger.js [hunk 3]
```diff
@@ -395,58 +393,60 @@ export default class Logger extends Writable {
...
 				host: this.host,
 				port: this.port
-			};			
+			};
 
-			this[$reconnection].on('connect', function (con) {
-				debugLogger.log("Connected");
+			this[$reconnection].on('connect', (connection) => {
+				this.debugLogger.log("Connected");
 				this.emit('connected');
```
Comment (gyakushev) [inline src/node_modules/logger.js]: are you sure `this` refers to correct object here?

File: src/node_modules/logger.js [hunk 4]
```diff
@@ -187,31 +191,35 @@ export default class Logger extends Writable {
...
 	@unwritable
-	async _write(log, enc, cb) {
-		log = finalizeLogString(log, this[$token]);
-
+	async _write(ch, enc, cb) {
 		try {
 			const conn = await this[$getConnection]();
-			conn.write(log);
+			var record = this.ringBuffer.read();
+			if (record != null) {
```
Comment (gyakushev) [inline src/node_modules/logger.js]: implementation below will return `undefined` on first `read()` call.

Comment (stopal-r7) [inline src/node_modules/logger.js]: `_write` is called whenever `write()` of `Writable` is called so there will always be some data in the array backing ringbuffer when we call `read()`.

File: src/node_modules/logger.js (hunk 3 above)
Comment (stopal-r7) [inline src/node_modules/logger.js]: yep, arrow function expression makes sure of that: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions

File: src/node_modules/logger.js (hunk 4 above)
Comment (gyakushev) [inline src/node_modules/logger.js]: Yes, that is correct. I still think a better version of this check is `if (record) {` to catch `undefined` value. It is difficult to imagine a scenario a `null` will be returned here.

Comment (StephenHynes7) [inline src/node_modules/logger.js]: `if (record)` may be slightly neater.

Comment (stopal-r7) [inline src/node_modules/logger.js]: Good point, that's what I was thinking, even though it is hard to imagine there will be an undefined/null return here, it is good to have such a check as a convenience.

File: src/node_modules/ringbuffer.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+export class RingBuffer extends EventEmitter {
+	constructor(opts) {
+		super();
+		this.records = [];
+		this.limit   = opts.bufferSize;
+	}
+
+	write(log) {
+		this.records.push(log);
+		if (this.records.length > this.limit) {
```
Comment (markstos) [inline src/node_modules/ringbuffer.js]: Firing an event here is recommended, so interested parties can be aware.  Currently the ring buffer creates a new "silent failure" case where logs start getting discarded, but no flags are raised to notify anyone.    

If an event were emitted, an listener could take action to email or or page someone, etc, to let someone know that logs have started being discarded.

The silent failure itself could be a security issue, as attacker aware of the design could to work to generate enough of a volume of logging to cause incriminating logs to be silently discarded.

Comment (stopal-r7) [inline src/node_modules/ringbuffer.js]: Good points @markstos, thanks for the review. 
We are actually emitting an event [here](https://github.com/rapid7/le_node/pull/122/files#diff-13bd19716a61073644f0cf6699227d6fR15) and we are listening to this in the Logger object that it belongs to. Then we log this event in internal logger of le_node once we start discarding events. Are you suggesting to emit another event for this in the Logger object itself here?

Comment (markstos) [inline src/node_modules/ringbuffer.js]: Ah, I missed that event. A single event will do. The only bit that I see missing then is document that it exists.  I would have the docs also recommend that developers use the event as a best practice.

Comment (stopal-r7) [inline src/node_modules/ringbuffer.js]: Updated the doc.

**Comment by danilosterrapid7** [comment]
+1
Just review the initialization of buffer size first... If it is that what you want.

**Pull request merged**